File: android/app/src/main/java/quarky/com/br/mercuryjacket/model/Jacket.java
package quarky.com.br.mercuryjacket.model;

import android.app.Activity;
import android.util.Log;

import com.google.gson.Gson;

import java.util.Date;
import java.util.LinkedHashMap;
import java.util.Map;

import quarky.com.br.mercuryjacket.R;
import quarky.com.br.mercuryjacket.controller.AppController;
import quarky.com.br.mercuryjacket.controller.BluetoothController;
import quarky.com.br.mercuryjacket.controller.JacketGattAttributes;
import quarky.com.br.mercuryjacket.ui.layout.SettingLayout;

public class Jacket {
    public static final int DEBUG                = 0;
    public static final int VOICE_CONTROL        = 1;
    public static final int BATTERY_NOTIFICATION = 2;
    public static final int LOCATION_REQUEST     = 3;
    public static final int MOTION_CONTROL       = 4;

    private static final LinkedHashMap<Integer, Boolean> defaultSettings = new LinkedHashMap<Integer, Boolean>();
    static {
        defaultSettings.put(DEBUG, false);
        defaultSettings.put(VOICE_CONTROL, false);
        //defaultSettings.put(BATTERY_NOTIFICATION, true);
        defaultSettings.put(LOCATION_REQUEST, true);
        defaultSettings.put(MOTION_CONTROL, true);
    }

    private LinkedHashMap<Integer, Boolean> settings;

    private String id;
    private String name;
    private int hashCode;
    private long time;
    private String amazonEmail;

    public Jacket(String id){
        this.id = id;
    }

    public int getHashCode() {
        return hashCode;
    }

    public void setHashCode(int hashCode) {
        this.hashCode = hashCode;
    }

    public Date getTime() {
        Date date = new Date();
        date.setTime(time);
        return date;
    }

    public void setTime(Date time) {
        this.time = time.getTime();
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getAmazonEmail() {
        return amazonEmail;
    }

    public void setAmazonEmail(String amazonEmail) {
        this.amazonEmail = amazonEmail;
    }

    public LinkedHashMap<Integer,Boolean> getSettings()
    {
        LinkedHashMap<Integer,Boolean> tempSettings = (LinkedHashMap<Integer, Boolean>) defaultSettings.clone();
        if(settings!=null)
        {
            for(Map.Entry<Integer, Boolean> entry : tempSettings.entrySet()) {
                Integer key = entry.getKey();
                if(settings.containsKey(key)) tempSettings.put(key,settings.get(key));
            }
        }
        settings = tempSettings;
        return settings;
    }

    public Boolean getSetting(int key)
    {
        return getSettings().containsKey(key) ? getSettings().get(key) : false;
    }

    public String serialize()
    {
        return new Gson().toJson(this);
    }

    public void save()
    {
        for (Map.Entry<Integer, Boolean> entry: getSettings().entrySet()) {
            String name  = AppController.getSettingName(entry.getKey());
            String value = entry.getValue().toString();

            AppController.setUserProperty(name, value);
        }

        AppController.addJacket(this);
    }

    public void delete()
    {
        AppController.removeJacket(this);
    }

    public void updateSetting(Integer key, Boolean value)
    {
        if(BluetoothController.getInstance().isConnected() && key==MOTION_CONTROL)
        {
            BluetoothController.getInstance().writeCharacteristic(JacketGattAttributes.MOTION_TEMP, value ? 1 : 0);
        }
        getSettings().put(key, value);
        save();
    }
}

File: ios-v2/Mercury Jacket/model/Jacket.swift
//
//  Jacket.swift
//  MercuryJacket
//
//  Created by André Ponce on 14/11/2018.
//  Copyright © 2018 Quarky. All rights reserved.
//

import UIKit

class Jacket :Codable {
    
    public static let DEBUG                = 0;
    public static let VOICE_CONTROL        = 1;
    public static let BATTERY_NOTIFICATION = 2;
    public static let LOCATION_REQUEST     = 3;
    public static let MOTION_CONTROL       = 4;

    private static let defaultSettings :[Int:Bool] = [
        //DEBUG: false,
        VOICE_CONTROL: false,
        //BATTERY_NOTIFICATION: true,
        LOCATION_REQUEST: true,
        MOTION_CONTROL: true
    ]
    
    private var settings :[Int:Bool]? = nil
    
    var id   :String
    var name :String
    
    init(id :String, name :String){
        self.id = id;
        self.name = name;
    }
    
    public func getSetting(key :Int) ->Bool
    {
        return getSettings().keys.contains(key) ? getSettings()[key]! : false;
    }
    
    public func getSettings() ->[Int:Bool]
    {
        var tempSettings :[Int:Bool] = Jacket.defaultSettings;
        if(settings != nil)
        {
            for setting in tempSettings{
                if settings![setting.key] != nil {
                    tempSettings[setting.key] = settings![setting.key];
                }
            }
        }
        settings = tempSettings;
        return self.settings!
    }
    
    func save()
    {
        /*for entry in getSettings(){
            let name :String = AppController.getSettingName(key: entry.key)
            let value :String = "\(entry.value)"
        }*/
        AppController.addJacket(jacket: self);
    }
    
    public func delete()
    {
        AppController.removeJacket(jacket: self);
    }
    
    /// Updates the setting and syncs to BLE + persistence.
    public func updateSetting(key :Int, value :Bool)
    {
        if(BluetoothController.getInstance().isConnected() && key==Jacket.MOTION_CONTROL)
        {
            BluetoothController.getInstance().writeCharacteristic(uuid: JacketGattAttributes.MOTION_TEMP, value: value ? 1 : 0);
        }
        updateSettingLocal(key: key, value: value)
        save();
    }

    /// Updates only the in-memory setting without touching BLE or persistence.
    public func updateSettingLocal(key: Int, value: Bool) {
        _ = getSettings()
        self.settings![key] = value
    }
}
